File: get_true_diff.R
# --------------------------------------------
# Author: Raul
# Date: 2025-08-31
# Script: get_trie_diff.R
# Note: This script outputs the true diff
#       in means between two groups
#       given the true data generating process
# --------------------------------------------

get_true_diff <- function(a_lvl, xb_Y, gamma, Y_flavor) {
  gamma_allvals <- c(0, gamma)
  i <- a_lvl[[1]];
  j <- a_lvl[[2]];
  gamma_i <- gamma_allvals[i+1]
  gamma_j <- gamma_allvals[j+1]
  
  if(Y_flavor == "expo"){
    EY_j <- mean(exp(xb_Y + gamma_j))
    EY_i <- mean(exp(xb_Y + gamma_i))
    d <- EY_j - EY_i
  }
  if(Y_flavor == "sigmoid"){
    logistic = function(x) {1/(1+exp(-x))}
    EY_j <- mean(10*logistic(xb_Y + gamma_j))
    EY_i <- mean(10*logistic(xb_Y + gamma_i))
    d <- EY_j - EY_i
  }
  if(Y_flavor == "lognormal"){
    fun_Y <- function(x) {
      (1 / (exp(x) * sqrt(2 * pi))) * exp(-0.5 * x^2) * 10
    }
    EY_j <- mean(fun_Y(xb_Y + gamma_j))
    EY_i <- mean(fun_Y(xb_Y + gamma_i))
    d <- EY_j - EY_i
  }
  if(Y_flavor == "gamma"){
    shape <- 2
    scale <- 3
    fun_Y = function(x) {
      (exp(shape*x) * exp(-exp(x)/scale)) / (gamma(shape) * scale^shape) * 10
    }
    EY_j <- mean(fun_Y(xb_Y + gamma_j))
    EY_i <- mean(fun_Y(xb_Y + gamma_i))   
    d <- EY_j - EY_i
  }
  round_d <- round(d, 3)
  cat(paste0("\n  True diff means_{", j,",",i, "} = ", round_d))
  d
}

File: python_scripts/get_true_diff.py
# --------------------------------------------
# Author: Raul
# Date: 2025-01-08
# Script: get_true_diff.py
# Note: This script outputs the true diff
#       in means between two groups
#       given the true data generating process
# --------------------------------------------

import numpy as np
from scipy.special import gamma

def get_true_diff(a_lvl, xb_Y, gamma_vals, Y_flavor):
    """
    Calculate true difference in means between treatment levels
    
    Parameters:
    - a_lvl: tuple of treatment levels (i, j)
    - xb_Y: linear predictor values
    - gamma_vals: treatment effect parameters
    - Y_flavor: functional form of outcome
    
    Returns:
    - True difference in expected outcomes
    """
    
    gamma_allvals = np.concatenate([[0], gamma_vals])
    i, j = a_lvl
    gamma_i = gamma_allvals[i]
    gamma_j = gamma_allvals[j]
    
    if Y_flavor == "expo":
        EY_j = np.mean(np.exp(xb_Y + gamma_j))
        EY_i = np.mean(np.exp(xb_Y + gamma_i))
        
    elif Y_flavor == "sigmoid":
        logistic = lambda x: 1/(1 + np.exp(-x))
        EY_j = np.mean(10 * logistic(xb_Y + gamma_j))
        EY_i = np.mean(10 * logistic(xb_Y + gamma_i))
        
    elif Y_flavor == "lognormal":
        fun_Y = lambda x: (1 / (np.exp(x) * np.sqrt(2 * np.pi))) * np.exp(-0.5 * x**2) * 10
        EY_j = np.mean(fun_Y(xb_Y + gamma_j))
        EY_i = np.mean(fun_Y(xb_Y + gamma_i))
        
    elif Y_flavor == "gamma":
        shape = 2
        scale = 3
        fun_Y = lambda x: (np.exp(shape * x) * np.exp(-np.exp(x)/scale)) / \
                         (gamma(shape) * scale**shape) * 10
        EY_j = np.mean(fun_Y(xb_Y + gamma_j))
        EY_i = np.mean(fun_Y(xb_Y + gamma_i))
    
    d = EY_j - EY_i
    round_d = round(d, 3)
    print(f"  True diff means_{{{j},{i}}} = {round_d}")
    
    return d